### ISSUE TITLE
Manipulation of JSON document during compaction?

### ISSUE THREAD

**Issue by bshamblen** [digitalbazaar/jsonld.js#115]
In issue #96 @dlongley references a video describing the basic uses of expansion and compaction. In that video the example shows the conversion of two comment message documents into a common format for processing. In that example, both of the source documents have properties that are a one-to-one map (`name` and `commentText`).

I have read through both the JSON-LD and JSON-LD API specs and haven't been able to find any documentation on how to handle more complex situations, where the source document values are at a different level within the document, or multiple properties need to be combined into one to make up the desired output.

The features I've been looking for, but haven't been able to find, are:

* Move a property in the source document to another level in the output. For example, move a sub property value to the root level, move a root level value to a sub-property, or move a value from one sub property to another sub property.
* Merge two or more properties into one. For example, merge first name and last name into full name.
* Split one property into multiples. For example, split full name into first and last name.

#### Input Document From External Data Source (with context)
```javascript
{
  "@context": {
    "schema": "http://schema.org/",
    "first_name": "schema:givenName",
    "last_name": "schema:familyName",
    "streetNumber": "schema:streetAddress",
    "streetName": "schema:streetAddress",
    "city": "schema:addressLocality",
    "state": "schema:addressRegion",
    "zip": "schema:postalCode"
  },
  "first_name": "Bob",
  "last_name": "Smith",
  "streetNumber": "123",
  "streetName": "Main St",
  "city": "San Jose",
  "state": "CA",
  "zip": "95125"
}
```

#### Actual Output After Compaction
```javascript
{
  "http://schema.org/addressLocality": "San Jose",
  "http://schema.org/addressRegion": "CA",
  "http://schema.org/familyName": "Smith",
  "http://schema.org/givenName": "Bob",
  "http://schema.org/postalCode": "95125",
  "http://schema.org/streetAddress": [
    "Main St",
    "123"
  ]
}
```

#### Desired Output After Compaction
```javascript
{
  "http://schema.org/address": {
      "http://schema.org/streetAddress": "123 Main St",
      "http://schema.org/addressLocality": "San Jose",
      "http://schema.org/addressRegion": "CA",
      "http://schema.org/postalCode": "95125"
  },
  "http://schema.org/familyName": "Smith",
  "http://schema.org/givenName": "Bob",
  "http://schema.org/name": "Bob Smith"
}
```

Any feedback on the proper way to handle the manipulation of properties would be greatly appreciated. I'm willing to write the code to pre-process the documents to get them into the proper format, if there's no built in way to handle these use cases. I just figured that since these are rather basic data mapping functions there must be a way to handle them natively, I'm just missing it.

**Comment by dlongley** [comment]
Hi @bshamblen,

Sorry for the slow response, it's been a busy week. First, this is a question that may be better raised at [json-ld.org's github repo](https://github.com/json-ld/json-ld.org) as you're addressing the standard processing rules, not just how one particular processor (eg: jsonld.js) works.

In short, the features you're looking for aren't part of the standard processing algorithms: expansion, compaction, flattening, and they aren't part of the non-standard (but potentially will become standardized) framing algorithm.

> Move a property in the source document to another level in the output. For example, move a sub property value to the root level, move a root level value to a sub-property, or move a value from one sub property to another sub property.

To some extent you can do this with framing, but not in the way you're requesting. You can "reframe" data that is already modeled as a graph to appear in different places in a tree structure. However, here you're dealing with data that isn't modeled as a graph and you want to try to auto-manipulate it so it can be interpreted as such (eg: by moving properties from one object to another). For your case, however, there's nothing in jsonld.js or any other JSON-LD processor that I know of that can do that. It's more of a general programming problem that needs to be solved prior to using a processor.
  
> Merge two or more properties into one. For example, merge first name and last name into full name.

"Merge" can take on a variety of different meanings. For example, here you want it to mean "concatenation" (or concatenation plus a particular delimiter). JSON-LD does not support a special concatenation feature, and if it did, we'd have to specify a variety of different rules for it. It's better left to general programming. You can "merge" data in another way in JSON-LD, which is to specify that two properties map to the same global URI -- and they can then be merged underneath the same term during compaction.

> Split one property into multiples. For example, split full name into first and last name.

Again, "split" can take on a variety of different meanings. Here you want to split a string value. JSON-LD doesn't support such a feature, it's left to general programming. With JSON-LD, you can "split" different values of different types (eg: strings and booleans or URLs and integers) such that they appear underneath different terms post compaction.

JSON-LD's processing operations are based around its data model and using different ways to display the same information. They are not designed to manipulate particular values using various operations like string concatenation, arithmetic, etc. With respect to interpreting JSON as JSON-LD, we tried to cover as many common cases as we could in JSON-LD 1.0, but we couldn't get all of them. With the next iteration we will try to capture even more.

If you have any comments about how you think JSON-LD should work as a standard, please send them to the json-ld.org github repo because you'll get a wider audience there and any feature requests are more likely to make it into the discussion for the next version of the standard.

**Comment by bshamblen** [comment]
@dlongley Thanks for taking the time to provide such a detailed explanation. That's what I was afraid of. We're trying to create an interface where a non-technical person could map data from various API sources, and use the JSON-LD context as the sole mapping language. Unfortunately, when I started diving into the more complex use cases, like the ones described above, I began to realize it may not work with the existing functionality (which is great by the way). We'll most likely end up writing a wrapper to pre-process the data, to provide the transformation functionality that we need, before compacting it.

Thanks again!
